src/encoder.rs

```diff
@@ -1137,187 +1137,201 @@ pub fn save_block_motion<T: Pixel>(
-  }
-  if !skip && seq.enable_cdef {
-    cw.bc.cdef_coded = true;
-  }
-  cw.bc.cdef_coded
+pub struct BlockEncoder<'a, T: Pixel> {
+  pub seq: &'a Sequence,
+  pub fs: &'a mut FrameState<T>,
+  pub cw: &'a mut ContextWriter,
+  pub bsize: BlockSize,
+  pub bo: BlockOffset,
+  pub fi: &'a FrameInvariants<T>,
+  pub luma_mode: PredictionMode,
+  pub chroma_mode: PredictionMode,
+  pub ref_frames: [RefType; 2],
+  pub mvs: [MotionVector; 2],
+  pub cfl: CFLParams,
+  pub tx_size: TxSize,
+  pub tx_type: TxType,
+  pub mode_context: usize,
+  pub mv_stack: &'a [CandidateMV],
+  pub rdo_type: RDOType,
+  pub skip: bool,
+  pub for_rdo_use: bool,
 }
 
-pub fn encode_block_b<T: Pixel>(
-  fi: &FrameInvariants<T>, fs: &mut FrameState<T>,
-  cw: &mut ContextWriter, w: &mut dyn Writer,
-  luma_mode: PredictionMode, chroma_mode: PredictionMode,
-  ref_frames: [RefType; 2], mvs: [MotionVector; 2],
-  bsize: BlockSize, bo: BlockOffset, skip: bool,
-  cfl: CFLParams, tx_size: TxSize, tx_type: TxType,
-  mode_context: usize, mv_stack: &[CandidateMV],
-  rdo_type: RDOType, for_rdo_use: bool
-) -> i64 {
-  let is_inter = !luma_mode.is_intra();
-  if is_inter { assert!(luma_mode == chroma_mode); };
-  let sb_size = if fi.sequence.use_128x128_superblock {
-    BlockSize::BLOCK_128X128
-  } else {
-    BlockSize::BLOCK_64X64
-  };
-  let PlaneConfig { xdec, ydec, .. } = fs.input.planes[1].cfg;
-  if skip {
-    cw.bc.reset_skip_context(bo, bsize, xdec, ydec);
-  }
-  cw.bc.blocks.set_block_size(bo, bsize);
-  cw.bc.blocks.set_mode(bo, bsize, luma_mode);
-  cw.bc.blocks.set_tx_size(bo, bsize, tx_size);
-  cw.bc.blocks.set_ref_frames(bo, bsize, ref_frames);
-  cw.bc.blocks.set_motion_vectors(bo, bsize, mvs);
-
-  //write_q_deltas();
-  if cw.bc.code_deltas && fs.deblock.block_deltas_enabled && (bsize < sb_size || !skip) {
-    cw.write_block_deblock_deltas(w, bo, fs.deblock.block_delta_multi);
+impl<'a, T> BlockEncoder<'a, T> where T: Pixel {
+  pub fn encode_block_a<'w>(&mut self, w: &'w mut dyn Writer) -> bool {
+    self.cw.bc.blocks.set_skip(self.bo, self.bsize, self.skip);
+    if self.fs.segmentation.enabled && self.fs.segmentation.update_map && self.fs.segmentation.preskip {
+      self.cw.write_segmentation(w, self.bo, self.bsize, false, self.fs.segmentation.last_active_segid);
+    }
+    self.cw.write_skip(w, self.bo, self.skip);
+    if self.fs.segmentation.enabled && self.fs.segmentation.update_map && !self.fs.segmentation.preskip {
+      self.cw.write_segmentation(w, self.bo, self.bsize, self.skip, self.fs.segmentation.last_active_segid);
+    }
+    if !self.skip && self.seq.enable_cdef {
+      self.cw.bc.cdef_coded = true;
+    }
+    self.cw.bc.cdef_coded
   }
-  cw.bc.code_deltas = false;
 
-  if fi.frame_type == FrameType::INTER {
-    cw.write_is_inter(w, bo, is_inter);
-    if is_inter {
-      cw.fill_neighbours_ref_counts(bo);
-      cw.write_ref_frames(w, fi, bo);
+  pub fn encode_block_b<'w>(&mut self, w: &'w mut dyn Writer) -> i64 {
```

Comment:
An implicit lifetime for `'w` is equivalent:

```rust
pub fn encode_block_b(&mut self, w: &mut dyn Writer) -> i64 {
```

Reply:
Fixed